python/Ryu.py

```diff
@@ -273,31 +282,247 @@ def changeSwitchForwardingState(self, dpid=None, header=None, actions=None, comm
+                           'idle_timeout': int(idle_timeout),
+                           'pkt_count': 0,
+                           'header': header,
+                           'actions': of_actions,
+                           'priority': priority,
+                           'command': "ADD"}
+
+        if int(hard_timeout) != 0:
+            timeout = now + int(hard_timeout)
+            hard =        {'timeout': timeout,
+                           'dpid': dpid,
+                           'domain' : domain,
+                           'idle_timeout': int(hard_timeout),
+                           'pkt_count': 0,
+                           'header': header,
+                           'actions': of_actions,
+                           'priority': priority,
+                           'command': "ADD"}
+
         if(command == "ADD"):
             command = ofp.OFPFC_ADD
             flags = ofp.OFPFF_SEND_FLOW_REM
+            self.api.pushTimeouts(idle, hard)
+            self.api.pushFlows(dpid, domain, header, actions, priority)
         elif(command == "DELETE_STRICT"):
             command = ofp.OFPFC_DELETE_STRICT
+            self.api.remove_flow(dpid, domain, header, priority)
         else:
             command = -1
         #self.logger.error("Sending flow mod with command: " + str(command))
         #self.logger.error("Datpath: " + str(datapath))
-
-        mod = parser.OFPFlowMod( datapath     = datapath,
+        
+        mod = parser.OFPFlowMod( datapath     = dp,
                                  priority     = int(priority),
                                  match        = match,
                                  cookie       = 0,
+                                 buffer_id    = ofp.OFP_NO_BUFFER,
                                  command      = command,
                                  idle_timeout = int(idle_timeout),
                                  hard_timeout = int(hard_timeout),
                                  actions      = of_actions,
                                  flags        = flags)
+        return mod
 
-        if(datapath.is_active == True):
-            datapath.send_msg(mod)
+    def OF13_flow(self, dp=None, domain=None,header=None, actions=None, command=None, idle_timeout=None, hard_timeout=None, priority=1):
+        ofp      = dp.ofproto
+        parser   = dp.ofproto_parser
+        
+        obj = {}
+        obj['ipv6_src'] = None
+        obj['ipv6_dst'] = None
+        obj['ipv4_src'] = None
+        obj['ipv4_dst'] = None
+        obj['eth_type']  = None
+        
+        if(header.has_key('phys_port')):
+            obj['in_port'] = int(header['phys_port'])
+            header['in_port'] = int(header['phys_port'])
+            del header['phys_port']
+
+        if(header.has_key('nw_src')):
+            if(header['nw_src'].version == 4):
+                obj['ipv4_src'] = (str(header['nw_src'].ip), str(header['nw_src'].netmask))
+                obj['eth_type'] = 0x800
```

Comment:
can we use an existing ENUM or create a new one so that we don't have to know IPv4 ethernet type is 0x800 and IPv6 is 0x86dd?